File: app/src/main/java/ru/serjik/engine/gl/ShaderProgram.java
package ru.serjik.engine.gl;

import android.opengl.GLES20;
import ru.serjik.utils.SerjikLog;

public abstract class ShaderProgram {
    private int programHandle;
    private int vertexShaderHandle;
    private int fragmentShaderHandle;

    public ShaderProgram(String combinedSourceCode) {
        String[] shaders = combinedSourceCode.split("====");
        if (shaders.length < 2) {
            SerjikLog.log("Shader source missing '====' delimiter between vertex and fragment shaders");
            this.programHandle = 0;
            return;
        }
        this.programHandle = createProgram(shaders[0], shaders[1]);
    }

    private int compileShader(int shaderType, String source) {
        int shaderHandle = GLES20.glCreateShader(shaderType);
        if (shaderHandle != 0) {
            GLES20.glShaderSource(shaderHandle, source);
            GLES20.glCompileShader(shaderHandle);
            int[] isCompiled = new int[1];
            GLES20.glGetShaderiv(shaderHandle, GLES20.GL_COMPILE_STATUS, isCompiled, 0);
            if (isCompiled[0] == 0) {
                SerjikLog.log("Could not compile shader " + (shaderType == GLES20.GL_VERTEX_SHADER ? "vertex" : "fragment") + ":");
                SerjikLog.log(GLES20.glGetShaderInfoLog(shaderHandle));
                GLES20.glDeleteShader(shaderHandle);
                return 0;
            }
        } else {
            SerjikLog.log("Could not compile shader " + shaderType + ":");
        }
        return shaderHandle;
    }

    private int createProgram(String vertexSource, String fragmentSource) {
        this.vertexShaderHandle = compileShader(GLES20.GL_VERTEX_SHADER, vertexSource);
        if (this.vertexShaderHandle == 0) return 0;
        this.fragmentShaderHandle = compileShader(GLES20.GL_FRAGMENT_SHADER, fragmentSource);
        if (this.fragmentShaderHandle == 0) {
            GLES20.glDeleteShader(this.vertexShaderHandle);
            return 0;
        }
        int programHandle = GLES20.glCreateProgram();
        if (programHandle == 0) return 0;
        GLES20.glAttachShader(programHandle, this.vertexShaderHandle);
        GLError.check("glAttachVertexShader");
        GLES20.glAttachShader(programHandle, this.fragmentShaderHandle);
        GLError.check("glAttachFragmentShader");
        GLES20.glLinkProgram(programHandle);
        int[] linkStatus = new int[1];
        GLES20.glGetProgramiv(programHandle, GLES20.GL_LINK_STATUS, linkStatus, 0);
        if (linkStatus[0] == GLES20.GL_TRUE) return programHandle;
        SerjikLog.log("Could not link program: ");
        SerjikLog.log(GLES20.glGetProgramInfoLog(programHandle));
        GLES20.glDeleteProgram(programHandle);
        GLES20.glDeleteShader(this.fragmentShaderHandle);
        this.fragmentShaderHandle = 0;
        GLES20.glDeleteShader(this.vertexShaderHandle);
        this.vertexShaderHandle = 0;
        return 0;
    }

    public static void releaseCompiler() {
        GLES20.glReleaseShaderCompiler();
    }

    protected int getUniformLocation(String name) {
        return GLES20.glGetUniformLocation(this.programHandle, name);
    }

    public void use() {
        GLES20.glUseProgram(this.programHandle);
    }

    protected int getAttribLocation(String name) {
        return GLES20.glGetAttribLocation(this.programHandle, name);
    }

    /**
     * Releases the GL program and attached shader resources. Must be called on the GL thread.
     */
    public void release() {
        if (this.vertexShaderHandle != 0) {
            GLES20.glDeleteShader(this.vertexShaderHandle);
            this.vertexShaderHandle = 0;
        }
        if (this.fragmentShaderHandle != 0) {
            GLES20.glDeleteShader(this.fragmentShaderHandle);
            this.fragmentShaderHandle = 0;
        }
        if (this.programHandle != 0) {
            GLES20.glDeleteProgram(this.programHandle);
            this.programHandle = 0;
        }
    }
}

File: app/src/main/java/ru/serjik/engine/gl/ShaderProgram.kt
package ru.serjik.engine.gl

import android.opengl.GLES20
import ru.serjik.utils.SerjikLog

abstract class ShaderProgram(combinedSourceCode: String) {

    private var programHandle: Int
    private var vertexShaderHandle: Int = 0
    private var fragmentShaderHandle: Int = 0

    init {
        val shaders = combinedSourceCode.split("====")
        if (shaders.size < 2) {
            SerjikLog.log("Shader source missing '====' delimiter between vertex and fragment shaders")
            programHandle = 0
        } else {
            programHandle = createProgram(shaders[0], shaders[1])
        }
    }

    private fun compileShader(shaderType: Int, source: String): Int {
        val shaderHandle = GLES20.glCreateShader(shaderType)
        if (shaderHandle != 0) {
            GLES20.glShaderSource(shaderHandle, source)
            GLES20.glCompileShader(shaderHandle)
            val isCompiled = IntArray(1)
            GLES20.glGetShaderiv(shaderHandle, GLES20.GL_COMPILE_STATUS, isCompiled, 0)
            if (isCompiled[0] == 0) {
                val typeStr = if (shaderType == GLES20.GL_VERTEX_SHADER) "vertex" else "fragment"
                SerjikLog.log("Could not compile shader $typeStr:")
                SerjikLog.log(GLES20.glGetShaderInfoLog(shaderHandle))
                GLES20.glDeleteShader(shaderHandle)
                return 0
            }
        } else {
            SerjikLog.log("Could not compile shader $shaderType:")
        }
        return shaderHandle
    }

    private fun createProgram(vertexSource: String, fragmentSource: String): Int {
        vertexShaderHandle = compileShader(GLES20.GL_VERTEX_SHADER, vertexSource)
        if (vertexShaderHandle == 0) return 0

        fragmentShaderHandle = compileShader(GLES20.GL_FRAGMENT_SHADER, fragmentSource)
        if (fragmentShaderHandle == 0) {
            GLES20.glDeleteShader(vertexShaderHandle)
            return 0
        }

        val programHandle = GLES20.glCreateProgram()
        if (programHandle == 0) return 0

        GLES20.glAttachShader(programHandle, vertexShaderHandle)
        GLError.check("glAttachVertexShader")
        GLES20.glAttachShader(programHandle, fragmentShaderHandle)
        GLError.check("glAttachFragmentShader")
        GLES20.glLinkProgram(programHandle)

        val linkStatus = IntArray(1)
        GLES20.glGetProgramiv(programHandle, GLES20.GL_LINK_STATUS, linkStatus, 0)
        if (linkStatus[0] == GLES20.GL_TRUE) return programHandle

        SerjikLog.log("Could not link program: ")
        SerjikLog.log(GLES20.glGetProgramInfoLog(programHandle))
        GLES20.glDeleteProgram(programHandle)
        GLES20.glDeleteShader(fragmentShaderHandle)
        fragmentShaderHandle = 0
        GLES20.glDeleteShader(vertexShaderHandle)
        vertexShaderHandle = 0
        return 0
    }

    protected fun getUniformLocation(name: String): Int =
        GLES20.glGetUniformLocation(programHandle, name)

    fun use() {
        GLES20.glUseProgram(programHandle)
    }

    protected fun getAttribLocation(name: String): Int =
        GLES20.glGetAttribLocation(programHandle, name)

    /**
     * Releases the GL program and attached shader resources. Must be called on the GL thread.
     */
    fun release() {
        if (vertexShaderHandle != 0) {
            GLES20.glDeleteShader(vertexShaderHandle)
            vertexShaderHandle = 0
        }
        if (fragmentShaderHandle != 0) {
            GLES20.glDeleteShader(fragmentShaderHandle)
            fragmentShaderHandle = 0
        }
        if (programHandle != 0) {
            GLES20.glDeleteProgram(programHandle)
            programHandle = 0
        }
    }

    companion object {
        fun releaseCompiler() {
            GLES20.glReleaseShaderCompiler()
        }
    }
}
